releases-implementation/README.md

```diff
@@ -0,0 +1,415 @@
+
+## 4. Proposal
+
+This section defines how releases are implemented, where they are, how they are developed, tested and delivered.
+
+Then it compares the proposed solution to the current one that we have for our new KaaS product that is based on Cluster API.
+
+### 4.1. Implementation of releases
+
+This document proposes to continue using `giantswarm/releases` repository for creating and delivering releases, albeit in a simplified way when compared to the vintage releases. Release resources are used in a minimal way, and cluster-$provider apps continue to be used for deploying workload clusters.
+
+Briefly put, cluster-$provider apps are still deployed in almost exactly same way, with the following few differences:
+- In cluster-$provider app manifest, instead of specifying `.spec.version`, we MUST specify `release.giantswarm.io/version` label.
+- Information about app version, catalog and dependencies MUST BE obtained from the Release resource (during Helm rendering phase) when deploying a production cluster.
+- Information about app version, catalog and dependencies CAN be overridden with Helm values when deploying a test cluster in e2e tests.
+
+The core of this proposal is the idea to decouple the app and component versions from the cluster-$provider apps, which will then enable us to have a very scalable process for working with releases, where we can easily create and manage many releases across many providers and develop mechanisms to enforce business logic across all of those.
+
+We will start with showing a slightly simplified version of the Release resource, then explain how it is delivered and finally consumed.
+
+Other parts the `giantswarm/releases` repository that we CAN continue to use are:
+- release notes,
+- announcements,
+- unit tests for releases (for enforcing various rules about releases).
+
+`requests.yaml` file usage SHOULD be deprecated in favour of more scalable and automated process based on GitHub actions and Renovate.
+
+#### 4.1.1. Creating new release
+
+With the current Release CRD definition, we create a Release resource in the following way:
+- All apps, both those deployed as App resources and as HelmRelease resources, MUST be added to Release `.spec.apps`. For every app:
+	- we MUST specify app name,
+	- we MUST specify app version,
+	- we CAN OPTIONALLY specify the name of the catalog from which the app is installed (“default” by default), and
+	- we CAN OPTIONALLY specify the app’s dependencies.
+- Kubernetes version MUST BE specified as a `.spec.components` entry,
+- Flatcar version and image variant MUST BE specified as `.spec.components` entries,
+- cluster-$provider app version MUST BE specified as a `.spec.components` entry.
+
```

Comment:
What about versions of CAPx controllers? Does that have any bearing on a Release? E.g. does a specific release need to use a specific version of the CAPi and CAPA controllers as changes between versions could effect behaviour of how the WCs are created.

Reply:
The CAPI & CAPA controllers are not part of the release because we can only run one set on a management cluster.

Reply:
I thought the watch label allowed for running multiple on the same MC and each would deal with different resources?

Reply:
> What about versions of CAPx controllers? Does that have any bearing on a Release? 

No.

> E.g. does a specific release need to use a specific version of the CAPi and CAPA controllers as changes between versions could effect behaviour of how the WCs are created.

While changes in both CAPI and CAPA controllers should be always backward compatible, and you should be always able to use same manifests even with newer CAPI/CAPA controllers, it is true that with some quite newer CAPI/CAPA controller, even when you create a cluster with the same release, it might get created slightly differently compared to e.g. 3 months ago with older CAPI/CAPA controllers. This is true also today with cluster-$provider apps (for all providers), and it does not change with this RFC. IIRC this is very much intended, because we want to be able to update MC controllers continuously (mostly for the sake of fixes).

> The CAPI & CAPA controllers are not part of the release because we can only run one set on a management cluster.

Yes, exactly this.

> I thought the watch label allowed for running multiple on the same MC and each would deal with different resources?

It did, but IIRC when we implemented watch label, it was back then when we wanted to run CAPI/CAPx controllers on the existing vintage MC, so that vintage MCs can run CAPI/CAPx controllers and deploy CAPI WCs. AFAIK we have abandoned that idea, and migration assumes that WCs are migrated to new CAPI-only MCs.